### PR TITLE
Use late static binding and protected methods for easy class extension

### PR DESCRIPTION
**Pull request by thewhit**
(instead of self/private) to allow class inheritance (i.e., to easily write your own logging function or override methods)

For instance, I'd like to customize how failedValidationAction() handles validation failures. To do this currently I have to duplicate the code for `refreshToken`,  `authorizePost`, and `failedValidationAction` to avoid my extended class from using the parent methods.

### CODE REVIEW

File: libs/csrf/csrfprotector.php [hunk 1]
```diff
@@ -106,52 +106,52 @@ public static function init($length = null, $action = null)
...
+			 * Check locally for a config.php then check for
 			 * a config/csrf_config.php file in the root folder
 			 * for composer installations
 			 */
 			$standard_config_location = __DIR__ ."/../config.php";
 			$composer_config_location = __DIR__ ."/../../../../../config/csrf_config.php";
 
 			if (file_exists($standard_config_location)) {
-				self::$config = include($standard_config_location);
+				static::$config = include($standard_config_location);
```
Comment (mebjas) [inline libs/csrf/csrfprotector.php]: I'd like to learn the difference here.

Comment (thewhit) [inline libs/csrf/csrfprotector.php]: It's probably overkill to reference properties with late static binding. This is just an instance of sloppy replace all.

Comment (mebjas) [inline libs/csrf/csrfprotector.php]: Yeah, I think we should figure out what all should be inheritable and what not. There should always be some abstraction.

File: libs/csrf/csrfprotector.php [hunk 2]
```diff
@@ -36,14 +36,14 @@ class csrfProtector
 		 * flag for cross origin/same origin request
 		 * @var bool
 		 */
-		private static $isSameOrigin = true;
+		protected static $isSameOrigin = true;
```
Comment (mebjas) [inline libs/csrf/csrfprotector.php]: this should remain pvt

File: libs/csrf/csrfprotector.php [hunk 3]
```diff
@@ -36,14 +36,14 @@ class csrfProtector
...
 
 		/*
 		 * Variable: $isValidHTML
 		 * flag to check if output file is a valid HTML or not
 		 * @var bool
 		 */
-		private static $isValidHTML = false;
+		protected static $isValidHTML = false;
```
Comment (mebjas) [inline libs/csrf/csrfprotector.php]: used internally hence pvt

File: libs/csrf/csrfprotector.php (hunk 1 above)
Comment (mebjas) [inline libs/csrf/csrfprotector.php]: config is loaded from config file, hence no need to override it. Also if more config is needed changing the file would do.

Comment (thewhit) [inline libs/csrf/csrfprotector.php]: You're right. For my purposes, I think only `logCSRFattack` and `failedValidationAction` need to be extended. Let me test to make sure and then I can push the more limited update for your review.

Comment (mebjas) [inline libs/csrf/csrfprotector.php]: @thewhit 
That's cool. In fact it'd be amazing if you could also add a test where you inherit the library and perform these actions.

I'm closing this PR for now, could you create an issue for this for further discussion.

Comment (thewhit) [inline libs/csrf/csrfprotector.php]: @mebjas Yes, will do. Thanks.

**Comment by mebjas** [comment]
Thanks :)

**Pull request closed without merging**